Gstreamer Plugin: Add SVT-HEVC Gstreamer plugin

Yeongjin-Jeong opened the pull request:
Add SVT-HEVC plugin that runs on latest stable release v1.14.4 of Gstreamer.

tianjunwork commented:
@Yeongjin-Jeong Thank you very much for your effort for making the patch to use SVT-HEVC with GStreamer.  That facilitates the usability of SVT-HEVC with GStreamer users. Since SVT-HEVC is still under development, there will be new APIs being exposed, how do you plan to maintain the plugin?
There is also the internal effort to implement GStreamer plugins for SVT-HEVC/VP9/AV1, which is a standalone project(not a patch file apply to gst) installs *so to gst. They are planed to release to Github next week. There is no short plan to upstream to gst.
Also there is a trello page tracks the features being implemented. https://trello.com/b/0xqT4xeY/svt-hevc
Please check that before planning, in case of any waste of engineering time.
Thank you again to promote SVT into GStreamer! Let's work together do finalize the GSreamer plugin.

Yeongjin-Jeong commented:
@tianjunwork 
Thank you for your review. I did not know that there was such an internal plan, because I did not see the trello page. I expect your standalone project to be released next week. :)

If there are no other issues with this patch, I was thinking about working on new APIs in the gstreamer upstream. But if you think that you are unnecessary, I will close this PR and do not work in gst upstream. What do you think?

tianjunwork commented:
Hi @Yeongjin-Jeong, I verify your PR, it works well:). Thanks again for the contribution.

The format of your PR is a patch to gst which is well prepared to upstream this plugin to gst. While an independent gst plugin is a better form for end user who doesn't need to have all gst dependency.
So what I plan is your patch and the independent gst plugin project coexist in this repo.
At the same time, let's work on upstream your patch to gst. The independent gst plugin project stays here for a sometime based on community feedback. 
What do you think about this plan?
If you agree, we need to work on the interfaces together to make them same on both version.
FYI, I have to merge the independent gst plugin project **first** for some other dependency. 

Some comment of your patch:
1. folder name: svt -> svthevc
2. everywhere: svth265 -> svthevc, svth265enc -> svthevcenc
3. codeVpsSpsPps only works for the first key frame. There is no support for repeat_header yet.
4. meson build --reconfigure; ninja -C build uninstalled doesn't work for me.
   I use: ninja -C build/ update

Yeongjin-Jeong commented:
@tianjunwork 
I agree with your plan. :) I think we can get more feedback if we upload to gst upstream with this patch format. When the independent gst plugin project is merged, Let's co-work on how to modify the interface.

And thanks for your comments on this patch.

comment 1,3 : I'll fix soon and force push to this PR.
comment 2 : I used 'h265' instead of 'hevc', because all the elements of gstreamer were using only 'h265'. I'll modify it, but I don't know how to think about it in gst upstream.
comment 4 : "ninja -C build;" is missing before "ninja -C build uninstalled;". I'll fix soon too.

tianjunwork commented:
Thanks for your support:)
For comment 2, SVT-HEVC is the brand name. I think better to keep it this way.

Yeongjin-Jeong commented:
@tianjunwork 
I modified all you mentioned in the comments. And I rebased the latest master branch and forced push.
Now we need to make the interface the same, What parts of the interface should I modify?

tianjunwork commented:
> @tianjunwork
> I modified all you mentioned in the comments. And I rebased the latest master branch and forced push.
> Now we need to make the interface the same, What parts of the interface should I modify?

Thanks:). Let's work together on unifying the interface. I will review the interfaces of your/my implementation and come up some suggestion.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++      GST_DEBUG_FUNCPTR (gst_svthevc_enc_propose_allocation);
++
++  g_object_class_install_property (gobject_class, PROP_INSERT_VUI,
++      g_param_spec_boolean ("insert-vui", "Insert VUI",
```
I will add the same property in the other implementation.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
++
++  g_object_class_install_property (gobject_class, PROP_HIERARCHICAL_LEVEL,
++      g_param_spec_uint ("hierarchical-level", "Hierarchical Level",
```
x264 use this setting, I kept the same in the other implementation.
https://github.com/GStreamer/gst-plugins-ugly/blob/master/ext/x264/gstx264enc.c
 g_param_spec_boolean ("b-pyramid", "B-Pyramid"
what do you think? We use "b-pyramid"?

Yeongjin-Jeong replied:
I did the same as the property of the ffmpeg plugin, I think too it would be better to set it equal to gstreamer property name.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
++
++  g_object_class_install_property (gobject_class, PROP_LOOKAHEAD_DISTANCE,
++      g_param_spec_uint ("lookahead-depth", "Lookahead Depth",
```
just "lookahead" to simplify the spelling?

Yeongjin-Jeong replied:
I fixed this too.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
++
++  g_object_class_install_property (gobject_class, PROP_ENCODER_MODE,
++      g_param_spec_uint ("encoder-mode", "Encoder Mode",
```
x264 use "speed-preset", gst-svt use "speed" for encoder-mode, so how about we use "speed" here?

Yeongjin-Jeong replied:
I think too that would be better :)

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
++
++  g_object_class_install_property (gobject_class, PROP_RC_MODE,
++      g_param_spec_uint ("rc-mode", "Ratecontrol Mode",
```
"rc-mode" ==> just "rc"?
The principle is that, if x264 doesn't have the same, align with svt itself.

Yeongjin-Jeong replied:
I fixed this too.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
++
++  g_object_class_install_property (gobject_class, PROP_SCENE_CHANGE_DETECTION,
++      g_param_spec_boolean ("sc-detection", "Scene Change Detection",
```
"sc-detection" ==> "scd"

Yeongjin-Jeong replied:
I fixed this too.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++
++  g_object_class_install_property (gobject_class, PROP_BASE_LAYER_SWITCH_MODE,
++      g_param_spec_uint ("baselayer-mode", "Base Layer Switch Mode",
++          "Random Access Prediction Structure type setting: 0=Use B-frames in the base layer pointing to the same past picture, 1=Use P-frames in the base layerRefer to Appendix A.1",
```
remove "Refer to Appendix A.1"

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
++
++  g_object_class_install_property (gobject_class, PROP_KEY_INT_MAX,
++      g_param_spec_int ("key-int-max", "Max key frame",
```
g_param_spec_int ("key-int-max", "**Key-frame maximal interval**",  ==> easier to understand

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++
++  g_object_class_install_property (gobject_class, PROP_INTRA_REFRESH,
++      g_param_spec_uint ("intra-refresh", "Intra Refresh",
++          "Use Periodic Intra Refresh instead of IDR frames: 1=CRA (Open GOP), 2=IDR (Closed GOP)",
```
Remove "Use Periodic Intra Refresh instead of IDR frames: ", since in x264, it's a boolean.
"1=CRA (Open GOP), 2=IDR (Closed GOP)" should be fine.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2059 @@
++#define PROP_TUNE_DEFAULT                   1
++#define PROP_LATENCY_MODE_DEFAULT           0
++#define PROP_BASE_LAYER_SWITCH_MODE_DEFAULT 0
++#define PROP_BITRATE_DEFAULT                (2 * 1024)
```
The default bitrate is 7000kbits/s

Yeongjin-Jeong commented:
Hello @tianjunwork 
I resolved all of your reviews. And I have also modified the patch to separate the commits which depending on the new API since SVT-HEVC v1.3.0 (eq. tile row / col, color format)

tianjunwork commented:
Hi @Yeongjin-Jeong, to address comment, you can still work on your local fork branch, after you push to your remote fork, Github will automatically update this PR with your incremental changes. 
With force push, I can't review what is changed compared with your last commit. So it is never recommended to use force push with PR.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
++    GST_PAD_ALWAYS,
++    GST_STATIC_CAPS ("video/x-h265, "
++        "width = (int) [1, MAX], "
++        "height = (int) [1, MAX], " "framerate = (fraction) [0, MAX]"));
```
width max: 8192, height max: 4320
min should at least be a macro block size, that is 64.

Yeongjin-Jeong replied:
I fixed this.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
++    GST_STATIC_CAPS ("video/x-raw, "
++        "format = (string) I420, "
++        "width = (int) [1, MAX], "
++        "height = (int) [1, MAX], " "framerate = (fraction) [0, MAX]"));
```
width max: 8192, height max: 4320
min should at least be a macro block size, that is 64.

Yeongjin-Jeong replied:
I fixed this too.

tianjunwork commented on gstreamer-plugin/git-patch/README.txt:
```diff
@@ -0,0 +1,23 @@
+- git checkout 1.14
+- meson build
+- cd subprojects/gst-plugins-bad
+- git apply ../SVT-HEVC/gstreamer_plugin/0001-svth265enc-Add-svth265-encoder-element.patch
```
new folder: gstreamer-plugin/**git-patch**/

Yeongjin-Jeong replied:
I fixed this too.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
+ subdir('x265')
+ subdir('zbar')
++subdir('svthevc')
+diff --git a/ext/svthevc/Makefile.am b/ext/svthevc/Makefile.am
```
How about svthevcenc instead? In case there will be svthevcdec later.

Yeongjin-Jeong replied:
Ok, but when I finished work, can I squash that commit and initial commit after your review?

Yeongjin-Jeong replied:
I fixed this too.

tianjunwork commented on gstreamer-plugin/git-patch/README.txt:
```diff
@@ -0,0 +1,23 @@
+
+3. Verify
+>> Gstreamer is now built with svt-hevc, sample command line:
+gst-launch-1.0 videotestsrc ! svth265enc ! fakesink
```
Please update the name of svt encoder here.

Yeongjin-Jeong replied:
I fixed this too.

tianjunwork commented:
Hi @Yeongjin-Jeong , thanks again for the update:)
Somehow, with this version, svthevc can't be found by gst-inspect-1.0 or gst-launch-1.0.
I manually copy the so to usr/lib/x86_64-linux-gnu/ where all the plugins are located, since I have gst already on my system.
If I have gst already installed, how to run the version built in gst-build? Thx.

Yeongjin-Jeong commented:
> Hi @Yeongjin-Jeong, to address comment, you can still work on your local fork branch, after you push to your remote fork, Github will automatically update this PR with your incremental changes.
> With force push, I can't review what is changed compared with your last commit. So it is never recommended to use force push with PR.

I'll do that next time, sorry. :)

> Hi @Yeongjin-Jeong , thanks again for the update:)
> Somehow, with this version, svthevc can't be found by gst-inspect-1.0 or gst-launch-1.0.
> I manually copy the so to usr/lib/x86_64-linux-gnu/ where all the plugins are located, since I have gst already on my system.
> If I have gst already installed, how to run the version built in gst-build? Thx.

gst-build provide a 'uninstalled' python script that lets you enter an uninstalled development environment. You can get into with ```ninja -C build uninstalled```. I'll add this to the README.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
++        "stream-format = (string) byte-stream, "
++        "alignment = (string) au, "
++        "profile = (string) { main, main-10, main-4:4:4 }")
++    );
```
The Main 4:4:4 profile allows for a bit depth of 8-bits per sample with support for 4:0:0, 4:2:0, 4:2:2, and 4:4:4 chroma sampling. 
SVT doesn't fully support main-4:4:4 yet.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
++      g_param_spec_int ("key-int-max", "Key-frame maximal interval",
++          "Distance Between Intra Frame inserted: -1=no intra update. -2=auto",
++          -2, 255, PROP_KEY_INT_MAX_DEFAULT, G_PARAM_READWRITE));
++
```
Forgot "G_PARAM_READWRITE |" ? :)

Yeongjin-Jeong replied:
I missed that, sorry. I'll fix.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
++        "width = (int) [ 64, 8192 ], " "height = (int) [ 64, 4320 ], "
++        "stream-format = (string) byte-stream, "
++        "alignment = (string) au, "
++        "profile = (string) { main, main-10, main-4:4:4 }")
```
svt-hevc doesn't support 4:0:0 yet. So main-4:4:4 is not fully supported.

Yeongjin-Jeong replied:
But the svt encoder seems to force the MainREXT profile of output in EB_YUV444 format. 
```
SVT [Error]: Instance 1: The input profile is not correct, should be 4 or MainREXT for YUV422 or YUV444 cases
```
In this case, I think we need to specify main-4:4:4 in src caps to avoid negotiation fail.

tianjunwork replied:
Sure, I will add it. Thx.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
++          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
++
++  gst_element_class_set_static_metadata (element_class,
++      "svthevcenc", "Codec/Encoder/Video", "H265 Encoder",
```
Could you make "H265 Encoder" to:
"Scalable Video Technology for HEVC Encoder (SVT-HEVC Encoder)"

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2123 @@
++    GST_VERSION_MINOR,
++    svthevcenc,
++    "svt-hevc encoder based H265 plugins",
++    plugin_init, VERSION, "GPL", GST_PACKAGE_NAME, GST_PACKAGE_ORIGIN)
```
Could you update the plugin info?

#ifndef VERSION
#define VERSION "1.0"
#endif
#ifndef PACKAGE
#define PACKAGE "gstreamer-svt-hevc"
#endif
#ifndef PACKAGE_NAME
#define PACKAGE_NAME "SVT-HEVC Encoder plugin for GStreamer"
#endif
#ifndef GST_PACKAGE_ORIGIN
#define GST_PACKAGE_ORIGIN "https://github.com/OpenVisualCloud"
#endif

GST_PLUGIN_DEFINE (GST_VERSION_MAJOR,
    GST_VERSION_MINOR,
    svthevcenc,
    "Scalable Video Technology for HEVC Encoder (SVT-HEVC Encoder)",
    plugin_init, VERSION, "GPL", PACKAGE_NAME, GST_PACKAGE_ORIGIN)

Yeongjin-Jeong replied:
I think the PACKAGE_NAME and PACKAGE_ORIGIN are not the information of SVT-HEVC encoder plugin, it seems to be the information of the package to which the plugin belongs. (e.g. gst-pluings-bad). What do you think?

tianjunwork replied:
Agree.

tianjunwork commented:
Hi @Yeongjin-Jeong, I am sorry that I made some review comment again. They are some minor changes.
Could you also help to review #179? I tried to make two version unified. Thank you:)

Yeongjin-Jeong commented:
> Hi @Yeongjin-Jeong, I am sorry that I made some review comment again. They are some minor changes.
> Could you also help to review #179? I tried to make two version unified. Thank you:)

Thank you for every review. I will review to #179 :)

tianjunwork commented:
Hi @Yeongjin-Jeong, I tired to add profile/level to my implementation. I don't get the logic of gst to set profile/level. Before init, all the logic follow
gst_pad_get_allowed_caps (GST_VIDEO_ENCODER_SRC_PAD (encoder)); what's the purpose? 
And after that gst_svthevc_enc_set_level_tier_and_profile to get the real profile?
Thank you:)

Yeongjin-Jeong commented:
> Hi @Yeongjin-Jeong, I tired to add profile/level to my implementation. I don't get the logic of gst to set profile/level. Before init, all the logic follow
> gst_pad_get_allowed_caps (GST_VIDEO_ENCODER_SRC_PAD (encoder)); what's the purpose?
> And after that gst_svthevc_enc_set_level_tier_and_profile to get the real profile?
> Thank you:)

Hello @tianjunwork, 
gst_pad_get_allowed_caps () is a task to find out the caps which we can provide from caps that downstream wants. If we allow profile: { main, main-4:4:4 } in src caps and downstream wants main-4:4:4 profile, gst_pad_get_allowed_caps() provides caps with only main-4:4:4.
``` (e.g. gst-launch-1.0 videotestsrc ! svthevcenc ! video/x-h265,profile=main-4:4:4 ! fakesink) ```
And the real profile of the encoder found in gst_svthevc_enc_set_level_tier_and_profile() is need for downstream negotiation.

If there are multiple profiles in allowed caps, we have to specify the profile depending on input pixel format. Does the svt-hevc encoder support profile auto selecting? If it does not, it seems to have to add that logic to gst_svthevc_enc_set_format().

And getcaps () that provides the our allowed input caps should be modified too. I'll add soon.

tianjunwork commented:
@Yeongjin-Jeong Thank you very much for the explanation. 
svt-hevc encoder doesn't support auto selecting for profile. Level/tier is calculated internally.
Sure, let's add proper profile negotiation, then both PR are ready to merge.

tianjunwork commented on gstreamer-plugin/git-patch/0001-svthevcenc-Add-svthevc-encoder-element.patch:
```diff
@@ -0,0 +1,2122 @@
++      }
++    }
++
++    level = gst_structure_get_string (s, "level");
```
I don't think you will get level and tier here, since they are not part of src_factory.
They will always be NULL, even if you specify: video/x-h265,profile=main-4:4:4,level=3.1

Yeongjin-Jeong replied:
I am not sure, if there is no specify field in src_factory (e.g. level), it will be provided in the allowed caps because it does not intersect caps. I'll check this.

tianjunwork commented:
Hi @Yeongjin-Jeong, you can also add I422_10LE in sink pad. And set encoderColorFormat accordingly.
https://github.com/OpenVisualCloud/SVT-HEVC/pull/179/commits/adda068d303a7c0d7d5712d620c5ec137ae40258

Yeongjin-Jeong commented:
Hello @tianjunwork, 
I added some patches to ensure restricts output profile and input formats, and that supporting 422 format.

tianjunwork commented:
Hi @Yeongjin-Jeong, I updated my PR accordingly. And added main-444-10 profile which is already supported by encoder lib.
Would you like to add it too? If you want to add it later, your PR is good to go and be merged.
Thank you for this big contribution!

Yeongjin-Jeong commented:
> Hi @Yeongjin-Jeong, I updated my PR accordingly. And added main-444-10 profile which is already supported by encoder lib.
> Would you like to add it too? If you want to add it later, your PR is good to go and be merged.
> Thank you for this big contribution!

Can SVT encoder support the range extension profile for main-4:4:4-10 in the header buffer? The gstreamer's h265parser parsed header as main-4:4:4, not main-4:4:4-10. 

So I'll add it later, because it will fail negotiation with main-4: 4: 4-10 whether this issue was in svt or gstreamer h265parser. :)

tianjunwork commented:
I don't quite follow your concern of h265parser. How do you know "h265parser parsed header as main-4:4:4, not main-4:4:4-10"? 
Below command line doesn't show any error for me. Could you send the command that fails?
`gst-launch-1.0 videotestsrc ! video/x-raw,format=Y444_10LE,width=100,height=100 ! svthevcenc ! h265parse ! avdec_h265 ! filesink location=jun.dat`
According to https://hevc.hhi.fraunhofer.de/HM-doc/namespace_profile.html
MAINREXT = 4 which indicates 21 range extensions profiles. So h265parser can't tell which one it is, and it doesn't need to know that to perform parsing work.

The next release of SVT-HEVC 1.4.0 is July. Would you like to wait after that to upstream to gstreamer to reduce the unnecessary rework?

Yeongjin-Jeong commented:
> According to https://hevc.hhi.fraunhofer.de/HM-doc/namespace_profile.html
> MAINREXT = 4 which indicates 21 range extensions profiles. So h265parser can't tell which one it is, and it doesn't need to know that to perform parsing work.

Yes, you're right. I have verified that profile_idc indicates MAINREXT. 
But h265parse checks which of the format range extensions profiles this profile indicated to, with using profile_tier_level in SPS as well as profile_idc, and parsed profile is sent downstream after set to field of caps. (See also : https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/blob/master/gst-libs/gst/codecparsers/gsth265parser.c#L2883)
So I guessed there is a problem about profile_tier_level in SVT or gstreamer h265parser.

> I don't quite follow your concern of h265parser. How do you know "h265parser parsed header as main-4:4:4, not main-4:4:4-10"?
> Below command line doesn't show any error for me. Could you send the command that fails?
> `gst-launch-1.0 videotestsrc ! video/x-raw,format=Y444_10LE,width=100,height=100 ! svthevcenc ! h265parse ! avdec_h265 ! filesink location=jun.dat`

If you use following command with built version of this PR, you can see in the log that h265parse has sent the EVENT_CAPS to fakesink as "main-444" instead of "main-444-10", even though we configured that encoder profile to MAINREXT and encoder bit depth to 10.

```
$ GST_DEBUG_FILE=gst.log GST_DEBUG_NO_COLOR=1 GST_DEBUG=GST_EVENT:5 gst-launch-1.0 videotestsrc num-buffers=100 ! video/x-raw,format=Y444_10LE ! svthevcenc ! video/x-h265,profile=main-4:4:4 ! h265parse ! fakesink
$ cat gst.log |grep "have event type caps" | grep "fakesink" | grep "profile"
```

> The next release of SVT-HEVC 1.4.0 is July. Would you like to wait after that to upstream to gstreamer to reduce the unnecessary rework?

Okay, I agree. Finally I'll add some commits to rebase this PR to v1.16 of gstreamer (new stable version has been released recently).

tianjunwork commented:
Hi @Yeongjin-Jeong Your command here fails to negotiate the pipeline. main-444 doesn't accept Y444_10LE input.
`$ GST_DEBUG_FILE=gst.log GST_DEBUG_NO_COLOR=1 GST_DEBUG=GST_EVENT:5 gst-launch-1.0 videotestsrc num-buffers=100 ! video/x-raw,format=Y444_10LE ! svthevcenc ! video/x-h265,profile=main-4:4:4 ! h265parse ! fakesink
$ cat gst.log |grep "have event type caps" | grep "fakesink" | grep "profile"`

Instead, I run:
` GST_DEBUG_FILE=gst.log GST_DEBUG_NO_COLOR=1 GST_DEBUG="*:6" gst-launch-1.0 videotestsrc num-buffers=100 ! video/x-raw,format=Y444_10LE ! svthevcenc ! video/x-h265,**profile=main-4:4:4-10** ! h265parse ! fakesink`

In gst.log, h265parse shows:
`<fakesink0:sink> accept caps of video/x-h265, **profile=(string)main-4:4:4-10**, stream-                format=(string)byte-stream, alignment=(string)au, level=(string)2, tier=(string)main, width=(int)320, height=(int)240, pixel-aspect-ratio=(fraction)1/1, framerate=(fraction)30/1, parsed=(boolean)true`

I think both svt and h265parser has no issue. Please confirm:)

I suggest that we finish the work here and merge this PR. Then we work on upstream to gst. How about rebase on gst-plugin-ugly? What do you think?

Yeongjin-Jeong commented:
Hello @tianjunwork 

> I think both svt and h265parser has no issue. Please confirm:)

Okay, I'll check it again.

> I suggest that we finish the work here and merge this PR. Then we work on upstream to gst. How about rebase on gst-plugin-ugly? What do you think?

Currently, this patch will not be available for upstream because it can cause conflicts when merging into the master branch. If you want this patch to sync with upstream, what do you think about rebasing this PR to the latest master branch (1.17) instead of the stable version(1.14 or 1.16)?

And it is true that plugins with license issue are included in gst-plugins-ugly, but x265 plugin is in gst-plugins-bad even though license is GPL2. I guess that is because it is not stable yet. 
So what do you think about working in gst-plugins-bad and getting upstream feedback?

tianjunwork commented:
Sure, let's base on gst-plugins-bad and latest master branch (1.17).

Yeongjin-Jeong commented:
hello @tianjunwork,
I added new patch for rebasing to gstreamer 1.17.0. I verified that it works in the meson build and jhbuild environments. :)

tianjunwork commented on gstreamer-plugin/git-patch/README.txt:
```diff
@@ -8,11 +8,10 @@
 - git clone https://gitlab.freedesktop.org/gstreamer/gst-build.git
 - cd gst-build
-- git checkout 1.14
 - meson build
```
is "git checkout 1.17" needed? or at lease tell user which version to use(master or which specific).

Yeongjin-Jeong replied:
I fixed that. thank you :)

tianjunwork commented:
@Yeongjin-Jeong Thank you very much👍. Could you check the profile negotiation issue? After confirm/rework on this, I can get both PR merged:)

Yeongjin-Jeong commented:
hello @tianjunwork, I checked again.
> Instead, I run with main-4:4:4-10:
> ` GST_DEBUG_FILE=gst.log GST_DEBUG_NO_COLOR=1 GST_DEBUG="*:6" gst-launch-1.0 videotestsrc num-buffers=100 ! video/x-raw,format=Y444_10LE ! svthevcenc ! video/x-h265,profile=main-4:4:4-10 ! h265parse ! fakesink`

I added the main-4:4:4-10 patch to local branch and tried again, But the format range extensions profiles are still parsed to main-4:4:4 instead of main-4:4:4-10. 
Did you run in https://github.com/OpenVisualCloud/SVT-HEVC/pull/179 PR? The profile parser of the codec-utils (e.g. gst_codec_utils_h265_caps_set_level_tier_and_profile() ) used in the PR will not parse the format range extensions profiles. 
So the negotiation would have been successful because the profile string in caps will not has been replaced and the downstream profile string has been passed.

> In gst.log, h265parse shows:
> `<fakesink0:sink> accept caps of video/x-h265, profile=(string)main-4:4:4-10, stream- format=(string)byte-stream, alignment=(string)au, level=(string)2, tier=(string)main, width=(int)320, height=(int)240, pixel-aspect-ratio=(fraction)1/1, framerate=(fraction)30/1, parsed=(boolean)true`

What is the version of gst-plugins-bad? h265parse supports format range extensions profiles from v1.14. Previous versions not replace downstream profile strings. I used version 1.17.0, and in this h265parse "profile = main-444" was output.

tianjunwork commented:
Yes, I run with #179.
gstreamer1.0-plugins-bad is already the newest version (**1.8.3-1ubuntu0.2**).
Since #179 works with current gst installation, I will merge it.
The profile compatibility will be an on-going effort for both of us. If any issue shows up, we will fix it later.
